R/utils.R

```diff
@@ -261,35 +256,38 @@ filter_null <- function(x) {
   x[!vapply(x, is.null, logical(1))]
 }
 
-assert_valid_callbacks <- function(output, inputs, state, user_function) {
+assert_valid_callbacks <- function(output, params, func) {
+  inputs <- params[vapply(params, function(x) 'input' %in% attr(x, "class"), FUN.VALUE=logical(1))]
+  state <- params[vapply(params, function(x) 'state' %in% attr(x, "class"), FUN.VALUE=logical(1))]
```

Comment:
This is filtering `params` for either `input` or `state` objects, I gather. Seems to me there are two more things we need to test:
- there's nothing in `params` that's neither `input` NOR `state`
- the first `state` (if any) comes after the last `input`

Reply:
Yes, that's correct; I'll add a check to verify that `params` only contains elements with `class` of `input` or `state`.

Can you explain why `state` must follow `input`? I understand that this is the convention, but one of the side effects of your suggested interface change is that the order is now unimportant, since we're assigning attributes via the `input` and `state` wrapper functions.

Reply:
The `vapply` line here is essentially a filter, which returns `TRUE` if the list element's `class` attributes include `input` or `state`. One could use `sapply` instead, but that function is not type safe. `vapply` requires you to provide `FUN.VALUE`, which addresses that issue. 

In some instances, `vapply` is also faster than `sapply`.

Reply:
> * there's nothing in `params` that's neither `input` NOR `state`

fixed in 61feba58d2f60caf45b1468cfdd5fb28cca6a1e0

Reply:
> Can you explain why `state` must follow `input`?

Correct me if my lack of R experience is mangling this...

It would be possible to relax this constraint, but as it stands, we split it up into two lists internally, and provide these separately to `dash-renderer`. When `dash-renderer` then makes the request associated with this callback, it sends us two lists of values. We use these to call the callback:
```r
output_value <- do.call(callback, args = as.list(request$body$inputs, request$body$state))
```
So if I'm understanding it correctly, we'll pass these to the callback with all the inputs first, then the state values. If that doesn't match the order they had in `params`, we'll have to at that point reshuffle these values into the `params` order. So in the end, assuming I have that right, I guess it's your call: would it be easier to enforce the ordering of inputs then state, or reshuffle elements here? I'd be fine with it either way; the ordered version is closer to dash-py, the reshuffled is more flexible.

Reply:
> * the first `state` (if any) comes after the last `input`

fixed in 703215088f1170e1922b0a14ef22c63eee2e05d5

Reply:
It makes sense to me to ensure `input`/`state` ordering.

The current solution uses R's run-length encoding function to handle the case when `input` or `state` is preceded and followed  by `state` or `input`, respectively. I imagine there are much better approaches, but this does the job for now.

Reply:
Thanks 🎉
`rle` looks fine to me. I wouldn't worry about performance here, it's not a hot path.